### ISSUE TITLE
[clang++] do not emit missing braces warning if all initializers are missing the same number of braces

### ISSUE THREAD

**Issue by t3nsor** [llvm/llvm-project#111542]
Currently `-Wmissing-braces` is disabled in some useful cases (see [here](https://github.com/llvm/llvm-project/blob/main/clang/test/SemaCXX/aggregate-initialization.cpp#L155)).

However the current rules still emit a warning in cases such as
```
struct A {
    int x;
};
struct B {
    int y;
};
struct S : A, B {
};
int main() {
    S s = {1, 2};
}
```
This pattern is used to implement an aggregate tuple template (templating has been omitted from this example for the sake of simplicity) which is useful in cases where tuples must be able to contain immovable types such as in senders & receivers based execution libraries.

I would like to suggest a generalization of the current rules that would avoid the warning in the example above: if all initializers provided are missing the same number of enclosing pairs of braces, then don't emit the warning. So in the `std::array` case there would be no warning because an outer pair of braces is missing, while in the above tuple case each initializer would be missing an inner pair of braces. Note that under this rule there would no longer be a warning for the `AggregateAndEmpty` case in `aggregate-initialization.cpp` but maybe there is a refinement of the rule that would preserve the warning in that case.
